### ISSUE TITLE
median is not implemented

### ISSUE THREAD

**Issue by yukoba** [HIPS/autograd#139]
```autograd.numpy.median()``` is not implemented.

**Comment by wqp89324** [comment]
Is there any plan to implement autograd.numpy.median()? I'm working on a loss function hopefully robust to outliers, which requires median calculation. 

Thanks!

**Comment by dcasasor-purdue** [comment]
`np.median(x)` can be replaced by `np.sum(x) - np.max(x) - np.min(x)`. I recently applied this formulation and got autograd to get my jacobians.

**Comment by maedoc** [comment]
@dcasasor-purdue that's hard to fact check,
```python
>>> x = np.random.randn(100, 100)
>>> l, r = x.sum(axis=0) - x.max(axis=0) - x.min(axis=0), np.median(x,axis=0)
>>> np.mean((l-r)**2)                                                       
100.7067891474971
```

btw, I saw that [TFP implements a percentile function](https://www.tensorflow.org/probability/api_docs/python/tfp/stats/percentile) so I guess it's a matter of spelunking in their source to see how the gradient is defined.

sorry @duvenaud the relu comment didn't lead to enlightenment for me.

**Comment by duvenaud** [comment]
The gradient will be zero for all elements except for the median element (or the two elements averaged to get the median).  Here's pseudocode:
```
def median_vjp(a, g):
  N = len(a)
  index_of_median = ((N - 1) / 2) + 0.5
  grad_out = np.zeros_like(a)
  s = numpy.argsort(a)
  if is_odd(N):
    grad_out[s[int(index_of_median)]] = g
  else:
    grad_out[s[int(index_of_median + 0.5)]] = g / 2
    grad_out[s[int(index_of_median - 0.5)]] = g / 2
  return grad_out
```

If you want you can make it divide the gradient evenly in the case where there's more than one identical element of a that affects the median.

**Comment by dcasasor-purdue** [comment]
> 
> 
> @dcasasor-purdue that's hard to fact check,
> 
> ```python
> >>> x = np.random.randn(100, 100)
> >>> l, r = x.sum(axis=0) - x.max(axis=0) - x.min(axis=0), np.median(x,axis=0)
> >>> np.mean((l-r)**2)                                                       
> 100.7067891474971
> ```
> 
> btw, I saw that [TFP implements a percentile function](https://www.tensorflow.org/probability/api_docs/python/tfp/stats/percentile) so I guess it's a matter of spelunking in their source to see how the gradient is defined.
> 
> sorry @duvenaud the relu comment didn't lead to enlightenment for me.

@maedoc you are absolutely right. My particular implementation is actually an implementation of the discontinuous `mid` function, which takes only three arguments and selects the middle value. This is intended for piecewise continuous process modeling, rather than for a general implementation of the median function. Thanks for the clarification.
